File: add-ons/auth-flows/api-express/src/routes/auth-flows.ts
/**
 * Auth Flows Routes - Registration, Email Verification, Password Reset
 */
import { Router, Request, Response } from "express";
import { z } from "zod";
import crypto from "crypto";
import bcrypt from "bcryptjs";
import { db } from "../db";
import { sendVerificationEmail, sendPasswordResetEmail } from "../services/email";

const router = Router();

// Token expiry settings
const EMAIL_VERIFICATION_EXPIRY_HOURS = 24;
const PASSWORD_RESET_EXPIRY_HOURS = 1;

// Validation schemas
const registerSchema = z.object({
  name: z.string().min(2).max(100),
  email: z.string().email(),
  password: z.string().min(8).max(100),
});

const verifyEmailSchema = z.object({
  token: z.string(),
});

const forgotPasswordSchema = z.object({
  email: z.string().email(),
});

const resetPasswordSchema = z.object({
  token: z.string(),
  password: z.string().min(8).max(100),
});

/**
 * Generate a secure random token
 */
function generateToken(): string {
  return crypto.randomBytes(32).toString("base64url");
}

/**
 * POST /api/v1/auth/register
 * Register a new user and send verification email
 */
router.post("/register", async (req: Request, res: Response) => {
  try {
    const data = registerSchema.parse(req.body);

    // Check if email already exists
    const existingUser = await db.user.findUnique({
      where: { email: data.email },
    });

    if (existingUser) {
      return res.status(400).json({
        error: "An account with this email already exists",
      });
    }

    // Hash password
    const hashedPassword = await bcrypt.hash(data.password, 12);

    // Create user
    const user = await db.user.create({
      data: {
        email: data.email,
        username: data.email.split("@")[0],
        displayName: data.name,
        hashedPassword,
        isActive: false,
        isVerified: false,
      },
    });

    // Create verification token
    const token = generateToken();
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + EMAIL_VERIFICATION_EXPIRY_HOURS);

    await db.verificationToken.create({
      data: {
        userId: user.id,
        token,
        type: "email_verification",
        expiresAt,
      },
    });

    // Send verification email
    await sendVerificationEmail(user.email, user.displayName || user.email, token);

    return res.json({
      message: "Registration successful. Please check your email to verify your account.",
      userId: user.id,
    });
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.errors[0].message });
    }
    console.error("Registration error:", error);
    return res.status(500).json({ error: "Internal server error" });
  }
});

/**
 * POST /api/v1/auth/verify-email
 * Verify user's email address with token
 */
router.post("/verify-email", async (req: Request, res: Response) => {
  try {
    const data = verifyEmailSchema.parse(req.body);

    // Find token
    const verification = await db.verificationToken.findFirst({
      where: {
        token: data.token,
        type: "email_verification",
        usedAt: null,
      },
    });

    if (!verification) {
      return res.status(400).json({
        error: "Invalid or expired verification token",
      });
    }

    // Check expiry
    if (verification.expiresAt < new Date()) {
      return res.status(400).json({
        error: "Verification token has expired",
      });
    }

    // Update user and mark token as used
    await db.$transaction([
      db.user.update({
        where: { id: verification.userId },
        data: { isVerified: true, isActive: true },
      }),
      db.verificationToken.update({
        where: { id: verification.id },
        data: { usedAt: new Date() },
      }),
    ]);

    return res.json({
      message: "Email verified successfully. You can now log in.",
      verified: true,
    });
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.errors[0].message });
    }
    console.error("Email verification error:", error);
    return res.status(500).json({ error: "Internal server error" });
  }
});

/**
 * POST /api/v1/auth/resend-verification
 * Resend verification email
 */
router.post("/resend-verification", async (req: Request, res: Response) => {
  try {
    const data = forgotPasswordSchema.parse(req.body);

    // Generic response to prevent email enumeration
    const genericResponse = {
      message: "If an account exists, a verification email has been sent.",
    };

    // Find user
    const user = await db.user.findUnique({
      where: { email: data.email },
    });

    if (!user) {
      return res.json(genericResponse);
    }

    if (user.isVerified) {
      return res.json({
        message: "This email is already verified. You can log in.",
      });
    }

    // Invalidate old tokens
    await db.verificationToken.updateMany({
      where: {
        userId: user.id,
        type: "email_verification",
        usedAt: null,
      },
      data: { usedAt: new Date() },
    });

    // Create new token
    const token = generateToken();
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + EMAIL_VERIFICATION_EXPIRY_HOURS);

    await db.verificationToken.create({
      data: {
        userId: user.id,
        token,
        type: "email_verification",
        expiresAt,
      },
    });

    // Send email
    await sendVerificationEmail(user.email, user.displayName || user.email, token);

    return res.json(genericResponse);
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.errors[0].message });
    }
    console.error("Resend verification error:", error);
    return res.status(500).json({ error: "Internal server error" });
  }
});

/**
 * POST /api/v1/auth/forgot-password
 * Request password reset email
 */
router.post("/forgot-password", async (req: Request, res: Response) => {
  try {
    const data = forgotPasswordSchema.parse(req.body);

    // Generic response to prevent email enumeration
    const genericResponse = {
      message: "If an account exists with that email, a password reset link has been sent.",
    };

    // Find user
    const user = await db.user.findUnique({
      where: { email: data.email },
    });

    if (!user) {
      return res.json(genericResponse);
    }

    // Invalidate old reset tokens
    await db.verificationToken.updateMany({
      where: {
        userId: user.id,
        type: "password_reset",
        usedAt: null,
      },
      data: { usedAt: new Date() },
    });

    // Create new token
    const token = generateToken();
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + PASSWORD_RESET_EXPIRY_HOURS);

    await db.verificationToken.create({
      data: {
        userId: user.id,
        token,
        type: "password_reset",
        expiresAt,
      },
    });

    // Send email
    await sendPasswordResetEmail(user.email, user.displayName || user.email, token);

    return res.json(genericResponse);
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.errors[0].message });
    }
    console.error("Forgot password error:", error);
    return res.status(500).json({ error: "Internal server error" });
  }
});

/**
 * POST /api/v1/auth/reset-password
 * Reset password with token
 */
router.post("/reset-password", async (req: Request, res: Response) => {
  try {
    const data = resetPasswordSchema.parse(req.body);

    // Find token
    const verification = await db.verificationToken.findFirst({
      where: {
        token: data.token,
        type: "password_reset",
        usedAt: null,
      },
    });

    if (!verification) {
      return res.status(400).json({
        error: "Invalid or expired reset token",
      });
    }

    // Check expiry
    if (verification.expiresAt < new Date()) {
      return res.status(400).json({
        error: "Reset token has expired",
      });
    }

    // Hash new password
    const hashedPassword = await bcrypt.hash(data.password, 12);

    // Update password and mark token as used
    await db.$transaction([
      db.user.update({
        where: { id: verification.userId },
        data: { hashedPassword },
      }),
      db.verificationToken.update({
        where: { id: verification.id },
        data: { usedAt: new Date() },
      }),
    ]);

    return res.json({
      message: "Password reset successfully. You can now log in with your new password.",
      success: true,
    });
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.errors[0].message });
    }
    console.error("Password reset error:", error);
    return res.status(500).json({ error: "Internal server error" });
  }
});

export default router;

File: add-ons/auth-flows/api-fastapi/app/api/v1/routes/auth_flows.py
"""Auth Flows API Routes - Registration, Email Verification, Password Reset."""
from datetime import datetime, timedelta
import secrets
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import get_db
from app.core.security import get_password_hash, verify_password
from app.core.email import send_verification_email, send_password_reset_email
from app.db.models import User, VerificationToken
from app.schemas.auth_flows import (
    RegisterRequest,
    RegisterResponse,
    VerifyEmailRequest,
    VerifyEmailResponse,
    ForgotPasswordRequest,
    ForgotPasswordResponse,
    ResetPasswordRequest,
    ResetPasswordResponse,
    ResendVerificationRequest,
)

router = APIRouter()

# Token expiry settings (can be overridden via environment)
EMAIL_VERIFICATION_EXPIRY_HOURS = 24
PASSWORD_RESET_EXPIRY_HOURS = 1


def generate_token() -> str:
    """Generate a secure random token."""
    return secrets.token_urlsafe(32)


@router.post("/register", response_model=RegisterResponse)
async def register(
    request: RegisterRequest,
    db: AsyncSession = Depends(get_db),
):
    """Register a new user and send verification email."""
    # Check if email already exists
    result = await db.execute(select(User).where(User.email == request.email))
    existing_user = result.scalar_one_or_none()

    if existing_user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="An account with this email already exists",
        )

    # Create user
    user = User(
        email=request.email,
        username=request.email.split("@")[0],  # Default username from email
        display_name=request.name,
        hashed_password=get_password_hash(request.password),
        is_active=False,  # Not active until email verified
        is_verified=False,
    )
    db.add(user)
    await db.flush()

    # Create verification token
    token = generate_token()
    verification = VerificationToken(
        user_id=user.id,
        token=token,
        type="email_verification",
        expires_at=datetime.utcnow() + timedelta(hours=EMAIL_VERIFICATION_EXPIRY_HOURS),
    )
    db.add(verification)
    await db.commit()

    # Send verification email
    await send_verification_email(user.email, user.display_name or user.email, token)

    return RegisterResponse(
        message="Registration successful. Please check your email to verify your account.",
        user_id=user.id,
    )


@router.post("/verify-email", response_model=VerifyEmailResponse)
async def verify_email(
    request: VerifyEmailRequest,
    db: AsyncSession = Depends(get_db),
):
    """Verify user's email address with token."""
    # Find token
    result = await db.execute(
        select(VerificationToken).where(
            VerificationToken.token == request.token,
            VerificationToken.type == "email_verification",
            VerificationToken.used_at.is_(None),
        )
    )
    verification = result.scalar_one_or_none()

    if not verification:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid or expired verification token",
        )

    # Check expiry
    if verification.expires_at < datetime.utcnow():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Verification token has expired",
        )

    # Get user
    result = await db.execute(select(User).where(User.id == verification.user_id))
    user = result.scalar_one_or_none()

    if not user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User not found",
        )

    # Update user
    user.is_verified = True
    user.is_active = True

    # Mark token as used
    verification.used_at = datetime.utcnow()

    await db.commit()

    return VerifyEmailResponse(
        message="Email verified successfully. You can now log in.",
        verified=True,
    )


@router.post("/resend-verification", response_model=RegisterResponse)
async def resend_verification(
    request: ResendVerificationRequest,
    db: AsyncSession = Depends(get_db),
):
    """Resend verification email."""
    # Find user
    result = await db.execute(select(User).where(User.email == request.email))
    user = result.scalar_one_or_none()

    if not user:
        # Don't reveal if email exists
        return RegisterResponse(
            message="If an account exists, a verification email has been sent.",
        )

    if user.is_verified:
        return RegisterResponse(
            message="This email is already verified. You can log in.",
        )

    # Invalidate old tokens
    result = await db.execute(
        select(VerificationToken).where(
            VerificationToken.user_id == user.id,
            VerificationToken.type == "email_verification",
            VerificationToken.used_at.is_(None),
        )
    )
    old_tokens = result.scalars().all()
    for old_token in old_tokens:
        old_token.used_at = datetime.utcnow()

    # Create new token
    token = generate_token()
    verification = VerificationToken(
        user_id=user.id,
        token=token,
        type="email_verification",
        expires_at=datetime.utcnow() + timedelta(hours=EMAIL_VERIFICATION_EXPIRY_HOURS),
    )
    db.add(verification)
    await db.commit()

    # Send email
    await send_verification_email(user.email, user.display_name or user.email, token)

    return RegisterResponse(
        message="If an account exists, a verification email has been sent.",
    )


@router.post("/forgot-password", response_model=ForgotPasswordResponse)
async def forgot_password(
    request: ForgotPasswordRequest,
    db: AsyncSession = Depends(get_db),
):
    """Request password reset email."""
    # Find user
    result = await db.execute(select(User).where(User.email == request.email))
    user = result.scalar_one_or_none()

    # Always return success to prevent email enumeration
    response = ForgotPasswordResponse(
        message="If an account exists with that email, a password reset link has been sent.",
    )

    if not user:
        return response

    # Invalidate old reset tokens
    result = await db.execute(
        select(VerificationToken).where(
            VerificationToken.user_id == user.id,
            VerificationToken.type == "password_reset",
            VerificationToken.used_at.is_(None),
        )
    )
    old_tokens = result.scalars().all()
    for old_token in old_tokens:
        old_token.used_at = datetime.utcnow()

    # Create new token
    token = generate_token()
    verification = VerificationToken(
        user_id=user.id,
        token=token,
        type="password_reset",
        expires_at=datetime.utcnow() + timedelta(hours=PASSWORD_RESET_EXPIRY_HOURS),
    )
    db.add(verification)
    await db.commit()

    # Send email
    await send_password_reset_email(user.email, user.display_name or user.email, token)

    return response


@router.post("/reset-password", response_model=ResetPasswordResponse)
async def reset_password(
    request: ResetPasswordRequest,
    db: AsyncSession = Depends(get_db),
):
    """Reset password with token."""
    # Find token
    result = await db.execute(
        select(VerificationToken).where(
            VerificationToken.token == request.token,
            VerificationToken.type == "password_reset",
            VerificationToken.used_at.is_(None),
        )
    )
    verification = result.scalar_one_or_none()

    if not verification:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid or expired reset token",
        )

    # Check expiry
    if verification.expires_at < datetime.utcnow():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Reset token has expired",
        )

    # Get user
    result = await db.execute(select(User).where(User.id == verification.user_id))
    user = result.scalar_one_or_none()

    if not user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="User not found",
        )

    # Update password
    user.hashed_password = get_password_hash(request.password)

    # Mark token as used
    verification.used_at = datetime.utcnow()

    await db.commit()

    return ResetPasswordResponse(
        message="Password reset successfully. You can now log in with your new password.",
        success=True,
    )
